java/tools/src/main/java/org/apache/arrow/tools/FileToStream.java

```diff
@@ -38,11 +38,21 @@
   public static void convert(FileInputStream in, OutputStream out) throws IOException {
     BufferAllocator allocator = new RootAllocator(Integer.MAX_VALUE);
     try (ArrowFileReader reader = new ArrowFileReader(in.getChannel(), allocator)) {
-      try (ArrowStreamWriter writer = new ArrowStreamWriter(reader.getVectorSchemaRoot(), reader, out)) {
-        for (ArrowBlock block: reader.getRecordBlocks()) {
-          reader.loadRecordBatch(block);
-          writer.writeBatch();
+      VectorSchemaRoot root = reader.getVectorSchemaRoot();
+      // load the first batch before instantiating the writer so that we have any dictionaries
+      reader.loadNextBatch();
+      try (ArrowStreamWriter writer = new ArrowStreamWriter(root, reader, out)) {
+        writer.start();
+        while (true) {
+          int rowCount = reader.getVectorSchemaRoot().getRowCount();
+          if (rowCount == 0) {
+            break;
+          } else {
+            writer.writeBatch();
+            reader.loadNextBatch();
```

Comment:
probably a question for @julienledem -- at least in C++, a basic unit of currency is `std::vector<std::shared_ptr<RecordBatch>>`, so to be unable to accumulate record batches in some kind of collection (like a `std::deque` or `std::vector`), particularly in a concurrent context, and write them out, would be very limiting.

With this API, I am not sure how you would do concurrent reads and writes of record batches. I don't think there's anything to be done right now, but I think we should open JIRAs at least pointing out the potential problems